R/check_group_variation.R

```diff
@@ -0,0 +1,196 @@
+    select <- all.vars(select)
+  }
+  if (inherits(by, "formula")) {
+    by <- all.vars(by)
+  }
+
+  # sanity check
+  if (is.null(by)) {
+    insight::format_error("Please provide the group variable using `by`.")
+  }
+  if (!all(by %in% colnames(x))) {
+    insight::format_error(
+      "The variable(s) speciefied in `by` were not found in the data."
+    )
+  }
+
+  # select all, if not given
+  if (is.null(select)) {
+    select <- setdiff(colnames(x), by)
+  }
+
+  # create all combinations that should be checked
+  combinations <- expand.grid(
+    variable = select,
+    group = by,
+    stringsAsFactors = FALSE
+  )
+  combinations <- combinations[combinations$variable != combinations$group, ]
+  combinations$type <- NA_character_
+
+  # initialize lists
+  for (i in seq_len(nrow(combinations))) {
+    combinations[i, "type"] <- .check_nested(
+      x,
+      combinations[i, "group"],
+      combinations[i, "variable"],
+      tolerance = tolerance,
+      only_balanced = only_balanced
+    )
+  }
+
+  combinations <- datawizard::data_relocate(
+    combinations,
+    select = "group",
+    before = "variable"
+  )
+  class(combinations) <- c("check_group_variation", class(combinations))
+  combinations
+}
+
+
+#' @export
+print.check_group_variation <- function(x, ...) {
+  if (insight::n_unique(x$group) > 1L) {
+    cat(insight::export_table(
+      x,
+      caption = c("Check group variation", "blue"),
+      by = "group"
+    ))
+  } else {
+    x_new <- x
+    x_new$group <- NULL
+    cat(insight::export_table(
+      x_new,
+      caption = c(sprintf("Check %s variation", x$group[1]), "blue")
+    ))
+  }
+
+  return(invisible(x))
+}
+
+
+# utils -------------------------------------------------------------
+
+.check_nested <- function(data, by, predictor, ...) {
+  UseMethod(".check_nested", data[[predictor]])
+}
+
+
+.check_nested.numeric <- function(data, by, predictor, tolerance = 1e-05, ...) {
+  # demean predictor
+  d <- datawizard::demean(
+    data,
+    select = predictor,
+    by = by,
+    verbose = FALSE,
+    add_attributes = FALSE
+  )
+
+  # get new names
+  within_name <- paste0(predictor, "_within")
+  between_name <- paste0(predictor, "_between")
+
+  is_between <- var(d[[between_name]], na.rm = TRUE) > tolerance
+  is_within <- var(d[[within_name]], na.rm = TRUE) > tolerance
+  is_both <- is_between && is_within
+
+  if (is_both) return("both")
+  if (is_between) return("between")
+  if (is_within) return("within")
+  NULL
+}
+
+
+.check_nested.default <- function(data, by, predictor, only_balanced = TRUE, ...) {
+  group <- data[[by]]
+  variable <- data[[predictor]]
+
+  complete <- stats::complete.cases(group, variable)
+  group <- as.factor(group[complete])
+  variable <- variable[complete]
+
+  n_uniques <- tapply(variable, group, insight::n_unique)
+  is_between <- all(n_uniques == 1L)
+  if (is_between) return("between")
+
+  if (!insight::has_single_value(is_between)) return("both")
```

Comment:
`is_between` should always be `FALSE` here, right?
Why the check for `has_single_value()`?

Reply:
Yes, this is wrong. Fixed it.